lib/Conversion/TritonGPUToLLVM/ConvertLayoutOpToLLVM/SharedToDotOperandFMA.cpp

```diff
@@ -70,164 +32,229 @@ Value getStructFromValueTable(ArrayRef<Value> vals,
-
-  Value mContig = i32_val(sizePerThread[order[1]]);
-
-  // threadId in blocked layout
-  auto threadIds = getThreadIds(thread, shapePerCTATile, sizePerThread, order,
-                                rewriter, loc);
-  Value threadIdM = threadIds[0];
-
-  Value offA0 = isARow ? _0 : mul(threadIdM, mContig);
-  Value offA1 = isARow ? mul(threadIdM, mContig) : _0;
-  SmallVector<Value> aOff(aNumPtr);
-  for (int i = 0; i < aNumPtr; ++i) {
-    aOff[i] = add(mul(offA0, strideA0), mul(offA1, strideA1));
+/**
+ * @brief put elements from Value vec to appropriate indexes in opValues array
+ *
+ * This function maps elements of 3d sub-tensor in linear array.
+ * Axes are arranged in following order from fastest to slowest: [nonKdim, kDim,
+ * bDim]
+ */
+void storeValuesInLinearVector(PatternRewriter &rewriter, Location loc,
+                               SmallVector<Value> &opValues, Value vec,
+                               unsigned kElems, unsigned nonKElems,
+                               unsigned kIdx, unsigned nonKIdx, unsigned bIdx,
+                               int kDim, int nonKDim, int bDim, int fastDim) {
+  auto vecTy = cast<VectorType>(vec.getType());
+  auto vectorSize = vecTy.getNumElements();
+  auto elemTy = vecTy.getElementType();
+  for (int elem = 0; elem < vectorSize; ++elem) {
+    unsigned outIdx[3] = {};
+    outIdx[bDim] = bIdx;
+    outIdx[kDim] = kIdx;
+    outIdx[nonKDim] = nonKIdx;
+
+    outIdx[fastDim] += elem;
+    auto idx = outIdx[bDim] * kElems * nonKElems + outIdx[kDim] * nonKElems +
+               outIdx[nonKDim];
+    opValues[idx] = extract_element(elemTy, vec, i32_val(elem));
   }
-  auto elemTy = typeConverter->convertType(aTensorTy.getElementType());
-
-  Type ptrTy = ptr_ty(rewriter.getContext(), 3);
-  SmallVector<Value> aPtrs(aNumPtr);
-  for (int i = 0; i < aNumPtr; ++i)
-    aPtrs[i] = gep(ptrTy, elemTy, aSmem.base, aOff[i]);
-
-  SmallVector<Value> vas;
-
-  int mShapePerCTATile = getShapePerCTATileForMN(dLayout, true /*isM*/);
-  int mSizePerThread = getSizePerThreadForMN(dLayout, true /*isM*/);
-
-  for (unsigned k = 0; k < K; ++k)
-    for (unsigned m = 0; m < M; m += mShapePerCTATile)
-      for (unsigned mm = 0; mm < mSizePerThread; ++mm) {
-        Value offset =
-            add(mul(i32_val(m + mm), strideAM), mul(i32_val(k), strideAK));
-        Value pa = gep(ptrTy, elemTy, aPtrs[0], offset);
-        Value va = load(elemTy, pa);
-        vas.emplace_back(va);
-      }
-
-  return getStructFromValueTable(vas, rewriter, loc, typeConverter, elemTy);
 }
 
-Value loadBFMA(Value B, Value llB, BlockedEncodingAttr dLayout, Value thread,
-               Location loc, const LLVMTypeConverter *typeConverter,
-               ConversionPatternRewriter &rewriter) {
-  auto bTensorTy = cast<MemDescType>(B.getType());
-  auto bLayout = cast<SharedEncodingAttr>(bTensorTy.getEncoding());
-  auto bShapePerCTA = getShapePerCTA(bTensorTy);
+Value loadFMAOp(Value dotOp, Value llA, BlockedEncodingAttr dLayout,
```

Comment:
There are lots of magic indexing and calculation in this function in general. Can you provide more comments inside to make it easier for others to follow?